gpflow/expectations.py

```diff
@@ -364,57 +372,90 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
     Compute the expectation:
     expectation[n] = <Ka_{Z1, x_n} Kb_{x_n, Z2}>_p(x_n)
         - Ka_{.,.}, Kb_{.,.} :: RBF kernels
-    Ka and Kb as well as Z1 and Z2 can differ from each other, but this is supported
-    only if the Gaussian p is Diagonal (p.cov NxD) and Ka, Kb have disjoint active_dims
-    in which case the joint expectations simplify into a product of expectations
+    Ka and Kb as well as Z1 and Z2 can differ from each other.
 
-    :return: NxMxM
+    :return: Nxdim(Z1)xdim(Z2)
     """
     if kern1.on_separate_dims(kern2) and isinstance(p, DiagonalGaussian):  # no joint expectations required
         eKxz1 = expectation(p, (kern1, feat1))
         eKxz2 = expectation(p, (kern2, feat2))
         return eKxz1[:, :, None] * eKxz2[:, None, :]
 
-    if feat1 != feat2 or kern1 != kern2:
-        raise NotImplementedError("The expectation over two kernels has only an "
-                                  "analytical implementation if both kernels are equal.")
+    Ka, Kb = kern1, kern2
 
-    kern = kern1
-    feat = feat1
+    with params_as_tensors_for(Ka, feat1, Kb, feat2):
 
-    with params_as_tensors_for(kern), params_as_tensors_for(feat):
         # use only active dimensions
-        Xcov = kern._slice_cov(tf.matrix_diag(p.cov) if isinstance(p, DiagonalGaussian) else p.cov)
-        Z, Xmu = kern._slice(feat.Z, p.mu)
+        Xcov = Ka._slice_cov(tf.matrix_diag(p.cov) if isinstance(p, DiagonalGaussian) else p.cov)
+
+        Z1, Xmu = Ka._slice(feat1.Z, p.mu)
 
         N = tf.shape(Xmu)[0]
         D = tf.shape(Xmu)[1]
 
-        squared_lengthscales = kern.lengthscales ** 2. if kern.ARD \
-            else tf.zeros((D,), dtype=settings.tf_float) + kern.lengthscales ** 2.
+        def get_squared_length_scales(kern):
+            squared_lengthscales = kern.lengthscales ** 2. if kern.ARD \
+                        else tf.zeros((D,), dtype=settings.tf_float) + kern.lengthscales ** 2.
+            return squared_lengthscales
+
+        if Ka == Kb:
+            La = get_squared_length_scales(Ka)
+            Lb = La
+            LaLb = La * 0.5
+        else:
+            La, Lb = map(get_squared_length_scales, (Ka, Kb))
+            LaLb = La * Lb / (La + Lb)  # average length scale
 
-        sqrt_det_L = tf.reduce_prod(0.5 * squared_lengthscales) ** 0.5
-        C = tf.cholesky(0.5 * tf.matrix_diag(squared_lengthscales) + Xcov)  # NxDxD
+        sqrt_det_L = tf.reduce_prod(LaLb) ** 0.5
+        C = tf.cholesky(tf.matrix_diag(LaLb) + Xcov)  # NxDxD
         dets = sqrt_det_L / tf.exp(tf.reduce_sum(tf.log(tf.matrix_diag_part(C)), axis=1))  # N
 
+        def get_C_and_z_terms(Z, C=C):
```

Comment:
docstrings: what are these C and z terms?

Reply:
tbh, I think you need add some math clarification of what's going on here, _not only for this function_